backend/logic/courses/courses.js

```diff
@@ -145,9 +145,128 @@ var addCourseOfferings = async courseOfferings => {
   }
 };
 
+/**
+ * Adds a list of coreq into the database
+ * @author Steven Li
+ * @param {JSON} coreq
+ *        {
+ *          "ECSE 428": ["ECSE 321"],
+ *          "MATH 270": ["MATH 140", "MATH 240"]
+ *        }
+ * @returns true if insertion was successful
+ * @throws error if MySQL connection failed
+ *         invalid format coreq if JSON format is wrong
+ *         false if insertion failed
+ */
+var addCoreq = async coreq => {
+  await format.verifyCoreq(coreq);
+  let connection = await mysql.getNewConnection();
+  let query =
+    "INSERT INTO course_coreq (course_code, coreq_course_code) VALUES (?, ?);";
+  try {
+    await connection.beginTransaction();
+    for (let course in coreq) {
+      for (let i = 0; i < coreq[course].length; i++) {
+        await connection.query(query, [course, coreq[course][i]]);
+      }
+    }
+    await connection.commit();
+    connection.release();
+    return true;
+  } catch (error) {
+    console.error(error);
+    await connection.rollback();
+    connection.release();
+    throw new Error(false);
+  }
+};
+
+/**
+ * Adds a list of prereq into the database
+ * @author Steven Li
+ * @param {JSON} prereq
+ *        {
+ *          "ECSE 428": ["ECSE 321"],
+ *          "MATH 270": ["MATH 140", "MATH 240"]
+ *        }
+ * @returns true if insertion was successful
+ * @throws error if MySQL connection failed
+ *         invalid format prereq if JSON format is wrong
+ *         false if insertion failed
+ */
+var addPrereq = async prereq => {
+  await format.verifyPrereq(prereq);
+  let connection = await mysql.getNewConnection();
+  let query =
+    "INSERT INTO course_prereq (course_code, prereq_course_code) VALUES (?, ?);";
+  try {
+    await connection.beginTransaction();
+    for (let course in prereq) {
+      for (let i = 0; i < prereq[course].length; i++) {
+        await connection.query(query, [course, prereq[course][i]]);
+      }
+    }
+    await connection.commit();
+    connection.release();
```

Comment:
you can do `connection.release()` in finally clause, check my PR #42

Reply:
will finally even run if you're returning a value in the try block?

Reply:
yes similar to `finally` in Java, it will always execute